pkg/cachedinstances/cloudprovider/cached_cloud_provider.go

```diff
@@ -0,0 +1,261 @@
+	// When we mutate the cache, we hold the exclusive (write) lock to avoid concurrent reads.
+	// When we read from the cache from other goroutines in the Peek() method, we obtain the read lock.
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case toLookupC <- toLookupIP:
+			toLookupIP = gostatsd.UnknownIP // enable GC
+			toLookupC = nil                 // ip has been sent; if there is nothing to send, the case is disabled
+		case toReturnInfoC <- toReturnInfo:
+			toReturnInfo = gostatsd.InstanceInfo{} // enable GC
+			toReturnInfoC = nil                    // info has been sent; if there is nothing to send, the case is disabled
+		case info := <-ownInfoSource:
+			ch.handleInstanceInfo(info)
+		case t := <-refreshTicker.C:
+			ch.doRefresh(t)
+		case statser := <-ch.emitChan:
+			ch.emit(statser)
+		}
+		if toLookupC == nil && len(ch.toLookupIPs) > 0 {
+			last := len(ch.toLookupIPs) - 1
+			toLookupIP = ch.toLookupIPs[last]
+			ch.toLookupIPs[last] = gostatsd.UnknownIP // enable GC
+			ch.toLookupIPs = ch.toLookupIPs[:last]
+			toLookupC = ch.ipSinkSource
+		}
+		if toReturnInfoC == nil && len(ch.toReturnInfo) > 0 {
+			last := len(ch.toReturnInfo) - 1
+			toReturnInfo = ch.toReturnInfo[last]
+			ch.toReturnInfo[last] = gostatsd.InstanceInfo{} // enable GC
+			ch.toReturnInfo = ch.toReturnInfo[:last]
+			toReturnInfoC = ch.infoSinkSource
+		}
+	}
+}
+
+func (ch *CachedCloudProvider) Peek(ip gostatsd.IP) (*gostatsd.Instance, bool /*is a cache hit*/) {
+	ch.rw.RLock()
+	holder, existsInCache := ch.cache[ip]
+	ch.rw.RUnlock()
+	if !existsInCache {
+		return nil, false
+	}
+	holder.updateAccess()
+	return holder.instance, true // can be nil, true
+}
+
+func (ch *CachedCloudProvider) IpSink() chan<- gostatsd.IP {
+	return ch.ipSinkSource
+}
+
+func (ch *CachedCloudProvider) InfoSource() <-chan gostatsd.InstanceInfo {
+	return ch.infoSinkSource
+}
+
+func (ch *CachedCloudProvider) EstimatedTags() int {
+	return ch.cloudProvider.EstimatedTags()
+}
+
+func (ch *CachedCloudProvider) RunMetrics(ctx context.Context, statser stats.Statser) {
+	// All the channels are unbuffered, so no CSWs
+	flushed, unregister := statser.RegisterFlush()
+	defer unregister()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-flushed:
+			ch.scheduleEmit(ctx, statser)
+		}
+	}
+}
+
+// scheduleEmit is used to push a request to the main goroutine requesting metrics
+// be emitted.  This is done so we can skip atomic operations on most of our metric
+// counters.  In line with the flush notifier, it is fire and forget and won't block
+func (ch *CachedCloudProvider) scheduleEmit(ctx context.Context, statser stats.Statser) {
+	select {
+	case ch.emitChan <- statser:
+		// success
+	case <-ctx.Done():
+		// success-ish
+	default:
+		// at least we tried
+	}
+}
+
+func (ch *CachedCloudProvider) emit(statser stats.Statser) {
+	// regular
+	statser.Gauge("cloudprovider.cache_positive", float64(ch.statsCachePositive), nil)
```

Comment:
Oh, yeah, I'd like to discuss the metrics we emit as I want' paying too much attention to it while refactoring. Was too much to handle already so I might have disrupted things a bit.
I think not all active components emit stats properly. Some didn't have their `RunMetricsWithContext()`/etc called, some probably never get a non-null statser. An audit should take place. I tried to patch up things a bit, but haven't verified every single component emits stats properly.

Reply:
I'm sure @adampetrovic or @MovieStoreGuy will let us know if it broke anything :)

In terms of metric changes in this PR, I don't believe there's any loss of fidelity - there's the CP state (cache), and CP actions (actual CP work), and they're just split out, so it's all good.  Maybe for k8s, write the PR without metrics, and then raise an issue to refactor metrics later?

Once we have k8s, we can assess what metrics it can sensible generate, how they overlap with what we already have, and if we need new names / tags / whatever.